### ISSUE TITLE
[REVIEW]: GPUE: Graphics Processing Unit Gross-Pitaevskii Equation solver

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1037]
**Submitting author:** @leios (<a href="http://orcid.org/0000-0002-3243-8918">James Schloss</a>)
**Repository:** <a href="https://github.com/GPUE-group/GPUE" target ="_blank">https://github.com/GPUE-group/GPUE</a>
**Version:** v1.0
**Editor:** @labarba
**Reviewer:** @mgalloy, @markbasham
**Archive:** <a href="https://doi.org/10.5281/zenodo.2182729" target="_blank">10.5281/zenodo.2182729</a>

## Status

[![status](http://joss.theoj.org/papers/068f32aabaf1f741ce0b0e0bb21cb071/status.svg)](http://joss.theoj.org/papers/068f32aabaf1f741ce0b0e0bb21cb071)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/068f32aabaf1f741ce0b0e0bb21cb071"><img src="http://joss.theoj.org/papers/068f32aabaf1f741ce0b0e0bb21cb071/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/068f32aabaf1f741ce0b0e0bb21cb071/status.svg)](http://joss.theoj.org/papers/068f32aabaf1f741ce0b0e0bb21cb071)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/GPUE-group/GPUE" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mgalloy & @markbasham, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @labarba know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/GPUE-group/GPUE">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0)?
- [x] **Authorship:** Has the submitting author (@leios) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/GPUE-group/GPUE">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0)?
- [x] **Authorship:** Has the submitting author (@leios) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @mgalloy, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by labarba** [comment]
@mgalloy, @markbasham — Thank you for agreeing to review this submission for JOSS, we really appreciate your volunteer efforts to make this adventure in new-wave publishing a success!

This REVIEW issue thread is the main place for proceeding with the review, but you are encouraged to open new issues in the repository of the submission, as appropriate, and linking to them here, to keep track. Feel free to ask any questions you may have!

The [JOSS Review Process](https://joss.readthedocs.io/en/latest/reviewer_guidelines.html) is described in our documentation site.

**Comment by labarba** [comment]
@leios : I notice that some DOI links in your reference list are broken. Please fix your bibtex file, push, and re-compile using the command `@whedon generate pdf` in its own comment.

**Comment by leios** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by leios** [comment]
I believe the DOIs have been fixed. Working on the contribution guidelines soon.

Thanks!

**Comment by leios** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by markbasham** [comment]
Hi @labarba and @mgalloy Finally back from my conference and over jetlag, and I'm taking a look at this review.  I have had a look through the documentation, and Im happy with what I need to do, but how do i tick the checkboxes at the start of this ticket?

**Comment by mgalloy** [comment]
@markbasham Just make sure you are logged into GitHub and you should be able to check the boxes.

**Comment by markbasham** [comment]
@mgalloy thanks, all sorted now, should have read the instructions more clearly.

**Comment by markbasham** [comment]
@mgalloy I'm having trouble building it, but my system is a bit custom, have you managed?

**Comment by leios** [comment]
@markbasham What errors are you getting? This is obviously something we should look into, but I think the build instructions in the documentation should work for most linux systems.

**Comment by mgalloy** [comment]
I am also having problems on my Mac (unfortunately, the only system with an NVIDIA GPU right now). I can send a more detailed report this weekend.

**Comment by markbasham** [comment]
@leios the code all works for me, as do the example processes etc.  One thing i need to look at is the performance of the code, now it uses 84% of my GPU when running, so its pretty well making use of the system, but i don't have much else to base this on.  I have had a look at the documentation of GPELab but i cant see any runtimes.  do you know what the runtime for a process comparable to your documented example is?

**Comment by leios** [comment]
Thanks for submitting the building issue. I'll be working on that today.

For the benchmarks, I think running a simple test, [like this one](https://web.archive.org/web/20171120181431/https://peterwittek.com/gpe-comparison.html), would be a good start for simple split-op performance testing. That link was from an older build of GPUE, but we have not lost any functionality since then (I don't even think the flags have changed significantly from that build). It also provides GPElab scripts for the test.

As a note: the test there is for a simple harmonic oscillator in 2D. This is now a unit test that is run with `gpue -u`. A lot of the 2/3D vortex functionality is harder to benchmark. GPElab can do it, but the documentation is hard to come by.

**Comment by markbasham** [comment]
@leios thanks, thats perfect

**Comment by markbasham** [comment]
@leios final thing, is there no DOI for the final thesis in the paper? I couldent find anything obvious, but there are often university repos now for thesis, a link would do.

**Comment by leios** [comment]
I cannot seem to find a DOI, but there is a permanent link: http://hdl.handle.net/11250/2467021.

This can be added to the reference section, if desired.

**Comment by markbasham** [comment]
@leios sorry for the delay, sounds good to me.

**Comment by markbasham** [comment]
@leios if your happy to add that reference, that's the only thing I'm waiting on, if you have can you ask Whedon for a build.  Cheers,  Mark

**Comment by leios** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by leios** [comment]
I think that should update the reference. Thanks!

**Comment by leios** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by markbasham** [comment]
@labarba I am now happy with this review, good luck @leios

**Comment by labarba** [comment]
@mgalloy — 👋  How are you getting along with the review? Thanks!

**Comment by leios** [comment]
Thanks a bunch for the review @markbasham!

As for the current state of the review: There was some difficulty building the code on the latest MacOS release (more info found here: https://github.com/GPUE-group/GPUE/issues/14). This has been noted, but it doesn't seem like there is anything we can do at this stage until Apple and Nvidia developers start talking to each other again.

**Comment by mgalloy** [comment]
I have a lead on a machine I can use, I am waiting on a way to login to it.

**Comment by mgalloy** [comment]
Finally got access to the machine. Everything looks good. I have completed my review and have checked off all items on the checklist.

**Comment by labarba** [comment]
@leios : I now have some editorial suggestions to polish your paper.

Page 1.
[par. 1] 

- “many interesting physical phenomenon” >> phenomena
- “allow for new discoveries that directly mimic what can be seen in experiments and are thus highly valuable for fundamental research”>> hyperbole. If the simulations mimic experiments, it’s not really a “new discovery,” is it. Could you tone this down? E.g., what’s the difference between “valuable” and “highly valuable”?  You can say “Numerical simulations of BECs that directly mimic what can be seen in experiments are valuable in fundamental research in these areas.”
- “In practice, the dynamics of BEC systems can often be found” >> (declutter) The dynamics of BEC systems can be found … 
- “Though there are many methods to solve the GPE, one of the most straightforward…” >> One of the most straightforward of many methods for solving the GPE… (and you get rid of the syntactic expletive “there are”)
- “however, there are no generalized software packages available…” >> (make new sentence and) “No generalized software packages are available…” (again, get rid of syntactic expletive)
- “Even so, there are several…” >> Packages do exist for simulating BECs with other methods…

[par. 2] 
- “with an emphasis on” >> emphasizing
- “vortex turbulence in two dimensions” >> 2D turbulence is an oxymoron; perhaps say “vortex dynamics in two and three dimensions.”

Page 2.
- “highly configurable physical system parameters” >> what’s the difference between configurable and highly configurable? Perhaps delete “highly.”
- “GPUE additionally features a highly performant numerical solver implementation” >> GPUE’s high-performance numerical solver improves over other suites (Refs.)

**Comment by leios** [comment]
@whedon generate pdf

I believe this version of the paper addresses all these points.

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by leios** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by leios** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by leios** [comment]
I made a few more smallscale changes to the paper and I believe it is good to go. Thanks for all the help so far and let me know what else needs to be done.

**Comment by labarba** [comment]
This remains:

* “allow for new discoveries that directly mimic what can be seen in experiments and are thus highly valuable for fundamental research”>> hyperbole. If the simulations mimic experiments, it’s not really a “new discovery,” is it. Could you tone this down? E.g., what’s the difference between “valuable” and “highly valuable”? You can say “Numerical simulations of BECs that directly mimic what can be seen in experiments are valuable in fundamental research in these areas.”

**Comment by labarba** [comment]
Actually, looks like several of my editorial suggestions were ignored. Can you go over the items in my comment above? https://github.com/openjournals/joss-reviews/issues/1037#issuecomment-440960650

**Comment by leios** [comment]
Ah, my mistake. A lot of the changes seem to be in a separate branch.

**Comment by labarba** [comment]
Typo: “No generalized software packages are vailable…” >> available

**Comment by labarba** [comment]
"features a numerical solver with a greater performance than other available suites" >> This reads like an unsubstantiated performance claim. Is there justification somewhere for this? (in the tests, in the docs…)

**Comment by leios** [comment]
There is a citation for performance tests after the latter claim (https://peterwittek.com/gpe-comparison.html). I suppose I could put the actual numbers in there and say that we are 206x faster than GPELab, 5.5x faster than TS-MPI CPU (the fastest known CPU implementation), and 1.19 faster than TS-MPI GPU in real-time and 147x faster than GPELab, 3.48x faster than TS-MPI CPU, and 1.21x faster than TS-MPI GPU in imaginary time; however, these are all in the reference.

**Comment by leios** [comment]
It did. The previous paper said the following: `The above features enable highly configurable physical system parameters, and allow for the simulation of state-of-the-art system dynamics. GPUE additionally features a highly performant numerical solver implementation, with performance greater than other available suites.` By your request, this was modified to simply `GPUE additionally features a numerical solver with a greater performance than other available suites`

**Comment by labarba** [comment]
My editorial suggestion above, in https://github.com/openjournals/joss-reviews/issues/1037#issuecomment-440960650

* “GPUE additionally features a highly performant numerical solver implementation” >> GPUE’s high-performance numerical solver improves over other suites (Refs.)

This mild claim didn't raise alarm bells. But your version: 

> features a numerical solver with a greater performance than other available suites

… raised a question in my head: Is this a performance claim? Is it justified? Did the reviewers check that? Because the reviewers have an item to check off that says: "If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)"

I think that if you say "my software gives greater performance than others," readers will want evidence.  You can soften the language, or give a link or citation to evidence. I don't think you should add numbers in the paper now, because we'd have to go back to the reviewers

**Comment by leios** [comment]
Alright, great! The link is there already, so I'll fix the typo and regenerate the paper?

**Comment by leios** [comment]
@whedon generate pdf

> GPUE additionally features a numerical solver with a greater performance than other available suites

is now

> GPUE additionally features a numerical solver that improves on other available suites

and 

> No generalized software packages are vailable

is now

> No software packages are available

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01037/joss.01037/10.21105.joss.01037.pdf)

**Comment by labarba** [comment]
Please now make a deposit of your submission in an archival-quality service (e.g., Zenodo), and report here the archive DOI. Thanks!

**Comment by leios** [comment]
Thanks for the help with the editorial process!

This should be the appropriate DOI: https://doi.org/10.5281/zenodo.2182729

Here's the badge, in case you need it: [![DOI](https://zenodo.org/badge/79505596.svg)](https://zenodo.org/badge/latestdoi/79505596)

**Comment by labarba** [comment]
@whedon set 10.5281/zenodo.2182729 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.2182729" target="_blank">10.5281/zenodo.2182729</a> is the archive.

**Comment by labarba** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1037 with the following error: 

   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    15    0    15    0     0    254      0 --:--:-- --:--:-- --:--:--   258
sh: 65: Syntax error: redirection unexpected
Looks like we failed to compile the Crossref XML

**Comment by labarba** [comment]
@arfon S.O.S. ❗️

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/115

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/115, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
Fixed a small bug @labarba. This should be good to go now.

**Comment by arfon** [comment]
Actually, I've broken this in a more complex way attempting a fix. Will process manually on my end until I can put a proper fix in place.

**Comment by arfon** [comment]
@mgalloy, @markbasham - many thanks for your reviews here and to @labarba for editing this submission ✨ 

@leios - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.01037 :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01037/status.svg)](https://doi.org/10.21105/joss.01037)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01037">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01037/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01037/status.svg
   :target: https://doi.org/10.21105/joss.01037
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01037/status.svg)](https://doi.org/10.21105/joss.01037)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by labarba** [comment]
One thing I forgot to say,  @leios — the author list on the Zenodo archive doesn't match the author list of the paper. Zenodo grabs authors from commit history: you have to edit manually to match!

**Comment by arfon** [comment]
> Since there are only 2 authors, the et al. seems odd. It's not a big deal if it cannot be changed.

Sorry, this can't be easily changed.

**Comment by leios** [comment]
Alright, no problem. Thanks!
